src/librustc_trans/trans/adt.rs

```diff
@@ -1056,8 +1083,16 @@ pub fn num_args(r: &Repr, discr: Disr) -> usize {
         General(_, ref cases, dtor) => {
             cases[discr.0 as usize].fields.len() - 1 - (if dtor_active(dtor) { 1 } else { 0 })
         }
-        RawNullablePointer { nndiscr, ref nullfields, .. } => {
-            if discr == nndiscr { 1 } else { nullfields.len() }
+        RawForbiddenValue { payload_discr, ref unit_fields, .. } => {
+            if discr == payload_discr {
+                // By definition of `RawForbiddenValue`, the payload case
+                // has exactly one field.
+                1
+            } else {
+                // In the unit case, we may have any number of fields,
+                // including 0.
```

Comment:
(since you're already adding this comment, it might not hurt to note in the comment that this situation is not paradoxical -- i.e. all having >0 fields implies is that all such fields will themselves be zero-sized, right?)